### PR TITLE
Add support for Workers AI provider by CloudFlare

### PR DESCRIPTION
**Pull request by clun**
This PR introduces the different models implements by CloudFlare WorkersAI
https://developers.cloudflare.com/workers-ai/

And especially:
- [Text Embeddings](https://developers.cloudflare.com/workers-ai/models/#text-embeddings)
- [Text Generation](https://developers.cloudflare.com/workers-ai/models/#text-generation)
- [Text-to-Image](https://developers.cloudflare.com/workers-ai/models/#text-to-image)

### CODE REVIEW

File: langchain4j-workers-ai/pom.xml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>com.squareup.retrofit2</groupId>
+            <artifactId>retrofit</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.squareup.retrofit2</groupId>
+            <artifactId>converter-gson</artifactId>
```
Comment (langchain4j) [inline langchain4j-workers-ai/pom.xml]: minor: would be nice to use jackson instead

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@EnabledIfEnvironmentVariable(named = "WORKERS_AI_ACCOUNT_ID", matches = ".*")
+class WorkerAIEmbeddingModelIT {
+
+    static WorkersAiEmbeddingModel embeddingModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        embeddingModel = WorkersAiEmbeddingModel.builder()
+                .modelName(WorkersAiModelName.BAAI_EMBEDDING_BASE)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java]:
```suggestion
                .accountId(System.getenv("WORKERS_AI_ACCOUNT_ID"))
```
minor: to follow Workers AI naming  (please change all model classes)

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                .token(System.getenv("WORKERS_AI_API_KEY"))
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java]:
```suggestion
                .apiToken(System.getenv("WORKERS_AI_API_TOKEN"))
```
minor: to follow Workers AI naming  (please change all model classes)

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        Assertions.assertNotNull(out.content());
+    }
+
+    @Test
+    void generateEmbeddings() {
+        List<TextSegment> data = new ArrayList<>();
+        data.add(new TextSegment("Sentence1", new Metadata()));
+        data.add(new TextSegment("Sentence2", new Metadata()));
+        Response<List<Embedding>> out = embeddingModel.embedAll(data);
+        Assertions.assertNotNull(out.content());
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java:42]: worth checking that it contains 2 embeddings

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    static WorkersAiChatModel chatModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        chatModel = WorkersAiChatModel.builder()
+                .modelName(WorkersAiModelName.LLAMA2_7B_FULL)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERS_AI_API_KEY"))
+                .buildChatModel();
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java]:
```suggestion
                .build();
```

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    static WorkersAiEmbeddingModel embeddingModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        embeddingModel = WorkersAiEmbeddingModel.builder()
+                .modelName(WorkersAiModelName.BAAI_EMBEDDING_BASE)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERS_AI_API_KEY"))
+                .buildEmbeddingModel();
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIEmbeddingModelIT.java]:
```suggestion
                .build;
```

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    static WorkersAiLanguageModel languageModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        languageModel = WorkersAiChatModel.builder()
+                .modelName(WorkersAiModelName.LLAMA2_7B_FULL)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERS_AI_API_KEY"))
+                .buildLanguageModel();
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java]:
```suggestion
                .build();
```

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+@Disabled("Requires a Workers ai account")
+@EnabledIfEnvironmentVariable(named = "WORKERS_AI_API_KEY", matches = ".*")
+@EnabledIfEnvironmentVariable(named = "WORKERS_AI_ACCOUNT_ID", matches = ".*")
+class WorkerAILanguageModelIT {
+
+    static WorkersAiLanguageModel languageModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        languageModel = WorkersAiChatModel.builder()
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java]:
```suggestion
        languageModel = WorkersAiLanguageModel.builder()
```

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static dev.langchain4j.data.message.SystemMessage.systemMessage;
+import static dev.langchain4j.data.message.UserMessage.userMessage;
+import static dev.langchain4j.model.output.FinishReason.STOP;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Disabled("Requires a Workers ai account")
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java]:
```suggestion
```
Aren't `@EnabledIfEnvironmentVariable` not enough?
(please change in all tests)

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                .modelName(WorkersAiModelName.LLAMA2_7B_FULL)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERS_AI_API_KEY"))
+                .buildLanguageModel();
+    }
+    @Test
+    void generateText() {
+        Response<String> joke = languageModel.generate("Tell me jokeC cloud");
+        Assertions.assertNotNull(joke);
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAILanguageModelIT.java:26]: please add proper assertions

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAiImageModelIT.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+    static WorkersAiImageModel imageModel;
+
+    @BeforeAll
+    static void initializeModel() {
+        imageModel = WorkersAiImageModel.builder()
+                .modelName(WorkersAiModelName.STABLE_DIFFUSION_XL)
+                .accountIdentifier(System.getenv("WORKERS_AI_ACCOUNT_ID"))
+                .token(System.getenv("WORKERS_AI_API_KEY"))
+                .buildImageModel();
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAiImageModelIT.java]:
```suggestion
                .build();
```

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiModelName.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
+package dev.langchain4j.model.workersai;
+
+/**
+ * Group model name in a class to ease the usage.
+ * It has been preferred as an Enum for flexibility.
+ */
+@SuppressWarnings("unused")
+public class WorkersAiModelName {
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiModelName.java]: minor: would be nice to have separate enums for each model type, similar to `OpenAiChatModelName`/`OpenAiEmbeddingModelName`/etc

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+     */
+    private byte[] executeQuery(String prompt, Image image, Image mask) {
+        try {
+            // Mapping inbound
+            WorkersAiImageGenerationRequest imgReq = new WorkersAiImageGenerationRequest();
+            imgReq.setPrompt(prompt);
+            if (image != null) {
+                if (image.url() != null) {
+                    imgReq.setImage(getPixels(image.url().toURL()));
+                    System.out.println("Adding Image");
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java]:
```suggestion
```

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+                }
+            }
+            if (mask != null) {
+                if (mask.url() != null) {
+                    imgReq.setMask(getPixels(mask.url().toURL()));
+                    System.out.println("Adding mask");
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java]:
```suggestion
```

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+                }
+            }
+
+            retrofit2.Response<ResponseBody> response = workerAiClient
+                    .generateImage(imgReq, accountIdentifier, modelName)
+                    .execute();
+
+            System.out.println(response.code() + " " + response.message());
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java]:
```suggestion
```

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+                buffer.flush();
+                return buffer.toByteArray();
+            }
+            throw new IllegalStateException("An error occured while generating image.");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public int[] getPixels(URL imageUrl) throws Exception {
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java:241]: Is there any other way to do this?

**Review by ghost — COMMENTED** [review]
@clun thank you!

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java (hunk 16 above)
Comment (clun) [reply langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiImageModel.java:138]: this one no idea, it is not documented by cloudFlare

File: langchain4j-workers-ai/pom.xml (hunk 1 above)
Comment (clun) [reply langchain4j-workers-ai/pom.xml]: was following what has been done for Hugging Face. I can change.

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java (hunk 5 above)
Comment (clun) [reply langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java]: The same superclass is building all models here.

File: langchain4j-workers-ai/pom.xml (hunk 1 above)
Comment (langchain4j) [reply langchain4j-workers-ai/pom.xml]: Right, gson was previously used, now we are migrating to jackson

File: langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java (hunk 5 above)
Comment (langchain4j) [reply langchain4j-workers-ai/src/test/java/dev/langchain4j/model/workersai/WorkerAIChatModelIT.java]: I understand the reason for this, but is just not consistent with other integrations.
You could use lombok's `@Builder` if you don't want to rewrite a lot of code

**Comment by clun** [comment]
All conversations are resolved with 2 commits today

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiEmbeddingModel.java [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+            }
+            dev.langchain4j.model.workersai.client.WorkersAiEmbeddingResponse.EmbeddingResult res = retrofitResponse.body().getResult();
+            // Single Vector expected
+            if (res.getShape().get(0) != 1) {
+                throw new RuntimeException("Unexpected shape: " + res.getShape());
+            }
+            List<Float> embeddings = res.getData().get(0);
+            float[] floatArray = new float[embeddings.size()];
+            for (int i = 0; i < embeddings.size(); i++) {
+                floatArray[i] = embeddings.get(i); // Unboxing Float to float
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiEmbeddingModel.java:160]: minor: you could change the type from List<Float> to float[] in `res.getData()` class and avoid doing conversion manually

File: langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiEmbeddingModel.java [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+            throw new RuntimeException("Unexpected response: " + retrofitResponse);
+        }
+        WorkersAiEmbeddingResponse.EmbeddingResult res = retrofitResponse.body().getResult();
+
+        List<List<Float>> embeddings = res.getData();
+        List<Embedding> embeddingsList = new ArrayList<>();
+        for (List<Float> embedding : embeddings) {
+            float[] floatArray = new float[embedding.size()];
+            for (int i = 0; i < embedding.size(); i++) {
+                floatArray[i] = embedding.get(i); // Unboxing Float to float
```
Comment (langchain4j) [inline langchain4j-workers-ai/src/main/java/dev/langchain4j/model/workersai/WorkersAiEmbeddingModel.java:242]: same here

**Review by ghost — APPROVED** [review]
@clun thank you a lot!

**Pull request merged**